### ISSUE TITLE
NullReferenceException in InternalFireQueuedAsync()

### ISSUE THREAD

**Issue by FantasyTeddy** [dotnet-state-machine/stateless#280]
Our application crashes without any predictable pattern with the following exception message:
> System.NullReferenceException
  at Stateless.StateMachine\`2.<InternalFireQueuedAsync>d__19.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
  at Stateless.StateMachine\`2.<InternalFireAsync>d__18.MoveNext()

We are calling the `FireAsync` method in a "fire & forget" manner (we do not await it) and rely on the state machine to correctly handle the occurring transitions. Since we use the default `FiringMode.Queued` we expect being able to do that.
At first we expected an error with our configuration of the state machine, but after we double checked everything, we looked into the source code of the `InternalFireQueuedAsync` method and have a suspicion about what could be causing the problem: The internal `_eventQueue` is a simple `Queue` and not a `ConcurrentQueue` and maybe this could cause an issue if two different tasks try to enqueu/dequeue something at the same time?!

As a workaround we are currently using the `FiringMode.Immediate` and queuing the `FireAsync` calls to get the behavior we are looking for.

If you could look into this or tell me if I'm totally wrong, I would appreciate it.

Best regards
FantasyTeddy

**Comment by HenningNT** [comment]
Do you call FireAsync in the entry/exit actions? Are those awaited? I've always been sceptical to the async part of stateless, because I think it doesn't guarantee run-to-completion. But I'm not that familiar with async/await yet, so maybe I'm just not trusting myself with it.

Seeing that there are two refs to MoveNext() in the stacktrace leads me to thinking that an exception is either unhandled, or that the state machine is in another state than it should be...

**Comment by Maximusya** [comment]
Project's readme contains a note in [Async triggers](https://github.com/dotnet-state-machine/stateless#async-triggers) section:

> Note: while StateMachine may be used asynchronously, it remains single-threaded and may not be used concurrently by multiple threads.

`FiringMode.Queued` - as far as I remember - was introduced to handle triggers fired during transitions. Not for the async support.

**Comment by FantasyTeddy** [comment]
@HenningNT : Yes we call `FireAsync` in the entry/exit actions, but we do not await them. We wrongly assumed that the `FiringMode.Queued` allows us to use the `FireAsync` in a fire & forget manner.

@Maximusya : That's what I realized as well. We started to implement our own wrapper for an active state machine which should handle multiple threads just fine.

Thanks for taking your time to answer my questions, I will close this issue now because it is an usage problem instead of a framework problem.
